Fix some issues in tls_openseel and udp

JackLau1222 opened the pull request:
Hi Timo,

I've test this patchset and found some issues.

I create a simple DTLS client and server case here https://github.com/JackLau1222/openssl-dtls-bio-example/tree/master/ffmpeg_case

### This PR fix:
1. dtls_handshake can't return positive code when it still in progressing
2. udp server mode haven't dest_addr so we need set it through last_recv_addr

I submit these fixes to your GitHub repo because the udp fix depends on your patchset. 
So if you think these fixes are good, please help me by sending these fixes.

Best regards
Jack

BtbN commented on libavformat/udp.c:
```diff
@@ -1137,6 +1139,8 @@ static int udp_write(URLContext *h, const uint8_t *buf, int size)
 
     if (!s->is_connected) {
+        if (!s->dest_addr_len && !s->dest_addr.ss_family)
+            ff_udp_get_last_recv_addr(h, &s->dest_addr, &s->dest_addr_len);
```
This should probably be split into another commit, since it's a functional change of which I'm not sure if correct.
It's a behaviour change after all. Not sure if udp.c should make this decisions for its downstream user.

JackLau1222 replied:
split it already

BtbN commented on libavformat/udp.c:
```diff
@@ -1036,7 +1038,7 @@ static int udp_read(URLContext *h, uint8_t *buf, int size)
     UDPContext *s = h->priv_data;
     int ret;
-    socklen_t addr_len = sizeof(s->last_recv_addr);
+    s->last_recv_len = sizeof(s->last_recv_addr);
```
This means the size will always just be sizeof(struct sockaddr_storage), making it rather pointless to expose the size as an extra argument.
Either it's entirely unneccesary to expose the size at all, since one can just use sizeof(struct sockaddr_storage), or the actual size received needs to be passed and thus also exported from the receive thread.
At which point I'd probably introduce a small "header struct" instead of keeping to complexify the math done for the header size.

JackLau1222 replied:
the s->last_recv_len is not always just be sizeof(struct sockaddr_storage), this line just init.
The real value is set by https://github.com/BtbN/FFmpeg/pull/3/files#diff-0b9a2ec1baaf72443258876585e9b70f9cbf6aca990711d342afb6d3d2b625fcR1093

BtbN replied:
The majority of the time that line will not run though, since receiving is handled by the rx_worker, so the function returns long before that line.

JackLau1222 replied:
> The majority of the time that line will not run though, since receiving is handled by the rx_worker, so the function returns long before that line.

But this situation the whole last_recv_addr is nonsense because it also will not run, isn't?
I think maybe we should return the real length when last_recv_addr exist

rx_worker only runs if fifo_size is set.
At least in WHIP and dtls handshake didn't use it for now

JackLau1222 replied:
> The majority of the time that line will not run though, since receiving is handled by the rx_worker, so the function returns long before that line.

Okay I see. If we will pass the real length in that situation, maybe we need check the addr.ss_family type, it will add more code and unnecessary.

So I remove this patch that add "socklen_t last_recv_len"

Thanks for your review!

BtbN commented on libavformat/network.h:
```diff
@@ -338,7 +338,7 @@ int ff_connect_parallel(struct addrinfo *addrs, int timeout_ms_per_address,
                         int (*customize_fd)(void *, int, int), void *customize_ctx);
 
-void ff_udp_get_last_recv_addr(URLContext *h, struct sockaddr_storage *addr);
+void ff_udp_get_last_recv_addr(URLContext *h, struct sockaddr_storage *addr, socklen_t *addr_len);
```
Changing this API without also changing its users (only tls_schannel for now) will break compilation.

JackLau1222 replied:
Sorry about that, I haven't windows device nearby, so I haven't review the tls_schannel.c

will fix later

BtbN commented on libavformat/tls_openssl.c:
```diff
@@ -661,6 +662,24 @@ static int tls_close(URLContext *h)
+    av_log(NULL, AV_LOG_DEBUG ,"TLS: Trace %s, len=%u, cnt=%u, size=%u, hs=%u\n",
+        (incoming? "RECV":"SEND"), length, content_type, size, handshake_type);
+}
+
```
Probably should be AV_LOG_TRACE?
Also, what exactly is this? And is the stuff it parses out of the buffer documented somewhere? Is it stable?

JackLau1222 replied:
Good idea,

This function aims to log the tls/dtls send or receive data info like content type.

I use this function to debug so that I found the udp dest_addr haven't set because I don't see this log has been printed, it works well after I fix the udp issue.

This function refers to RFC 2246, 4346 5246(TLS 1.0, 1.1, 1.2),  RFC 4347 6347(DTLS 1.0, 1.2)


Will add these documents into comments

BtbN commented on libavformat/tls_openssl.c:
```diff
@@ -720,11 +720,10 @@ static int dtls_handshake(URLContext *h)
 
     r0 = SSL_do_handshake(p->ssl);
-    r1 = SSL_get_error(p->ssl, r0);
+    ret = r1 = SSL_get_error(p->ssl, r0);
```
Is it really correct to set ret like this?
Looking at the docs and headers, SSL_get_error sets positive error codes, so will never set this functions status to a failing one.

JackLau1222 replied:
> Is it really correct to set ret like this? Looking at the docs and headers, SSL_get_error sets positive error codes, so will never set this functions status to a failing one.


If the positive error code Is not `SSL_ERROR_WANT_READ`, `SSL_ERROR_WANT_WRITE`, `SSL_ERROR_ZERO_RETURN`

It will trigger the fail end:
https://github.com/BtbN/FFmpeg/pull/3/files#diff-4e72c01c8b4d085bcea7a05e60e953d7ac69d58a4e0b4563b7b7f2557c433f46R726

print_ssl_error will return a negative number like AVERROR(EIO)

BtbN replied:
Why do you set ret here then?
It will only ever be 0 or some positive OpenSSL error code here, which will then be returned if it was a non fatal error.
But the positive return code has no meaning outside of this function, and isn't use anywhere either from what I can tell.

JackLau1222 replied:
Firstly, why I set the positive return code is because avio_handshake needs `0 on a complete and successful handshake > 0 if the handshake progressed, but is not complete < 0 for an AVERROR code`

If the udp is in NONBLOCK mode, the handshake requires multi-steps(every time just finish one flight, there're four or six flights), The previous version return 0(success) when the first flight just finished, so it's wrong.
We need return positive code when the handshake is still processing.

Thanks for your reminder, maybe I should replace the positive OpenSSL error code to FFmpeg positive code like EINPROGRESS?

BtbN commented on libavformat/udp.c:
```diff
@@ -1144,6 +1144,10 @@ static int udp_write(URLContext *h, const uint8_t *buf, int size)
+        if (!s->dest_addr_len && !s->dest_addr.ss_family) {
+            ff_udp_get_last_recv_addr(h, &s->dest_addr);
+            s->dest_addr_len = sizeof(s->dest_addr);
+        }
```
I think this patch should be sent to the ML, since I'm very much not sure if this is a sensible thing to do.
It changes the behaviour of the udp URLContext in a potentially unexpected way.

Also, I added the addr length in my latest update of the patchset.

JackLau1222 replied:
Got it, I've update the code to match latest implementation.

Should I send my patch after you send the latest patch set? Or should I send my patch first to get others to review it?

Thanks

BtbN replied:
I'll apply the set soon, you can already send it though and just mention that it depends on my patchset.

JackLau1222 replied:
Thanks, I'll submit this patchset today, sry for late reply, I've been taking various final exams over the past few days.

JackLau1222 commented:
Hi Timo,

I noticed you have tried that test DTLS using ffmpeg command but meet error, I test in my local using such command(need base on this udp patch):
```shell
ffmpeg -v debug -listen 1 -cert_file cert.pem -key_file key.pem -mtu 1500 -autodetect_dest 1 -i dtls://127.0.0.1:4444 -c copy -y out.mp4

ffmpeg -v trace -f lavfi -re -i testsrc2=duration=10:size=128x72:rate=30 -an -sn -c:v libx264 -cert_file cert.pem -key_file key.pem -mtu 10000 -f mpegts "dtls://127.0.0.1:4444"
```
I modify this command because openssl seems not like too large MTU, so I decreased the resolution.
It seems work well on the part of dtls, but server can't pause reading from client so the data can't writing into mp4. the log like this:
```shell
> ffmpeg -v debug -listen 1 -cert_file cert.pem -key_file key.pem -mtu 1500 -autodetect_dest 1 -i dtls://127.0.0.1:4444 -c copy -y out.mp4
ffmpeg version N-120234-g9d0c964af2 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 17.0.0 (clang-1700.0.13.5)
  configuration: --enable-gpl --enable-shared --disable-static --disable-optimizations --enable-libx264 --enable-libx265 --enable-libopus --enable-sdl2 --enable-debug=3 --disable-stripping --enable-openssl --enable-version3 --extra-cflags='-fsanitize=address -g -O0' --extra-cxxflags='-fsanitize=address -g -O0' --extra-ldflags='-fsanitize=address -g -O0'
  libavutil      60.  4.101 / 60.  4.101
  libavcodec     62.  6.100 / 62.  6.100
  libavformat    62.  1.102 / 62.  1.102
  libavdevice    62.  0.100 / 62.  0.100
  libavfilter    11.  1.100 / 11.  1.100
  libswscale      9.  0.100 /  9.  0.100
  libswresample   6.  0.100 /  6.  0.100
Splitting the commandline.
Reading option '-v' ... matched as option 'v' (set logging level) with argument 'debug'.
Reading option '-listen' ... matched as AVOption 'listen' with argument '1'.
Reading option '-cert_file' ... matched as AVOption 'cert_file' with argument 'cert.pem'.
Reading option '-key_file' ... matched as AVOption 'key_file' with argument 'key.pem'.
Reading option '-mtu' ... matched as AVOption 'mtu' with argument '1500'.
Reading option '-autodetect_dest' ... matched as AVOption 'autodetect_dest' with argument '1'.
Reading option '-i' ... matched as input url with argument 'dtls://127.0.0.1:4444'.
Reading option '-c' ... matched as option 'c' (select encoder/decoder ('copy' to copy stream without reencoding)) with argument 'copy'.
Reading option '-y' ... matched as option 'y' (overwrite output files) with argument '1'.
Reading option 'out.mp4' ... matched as output url.
Finished splitting the commandline.
Parsing a group of options: global .
Applying option v (set logging level) with argument debug.
Applying option y (overwrite output files) with argument 1.
Successfully parsed a group of options.
Parsing a group of options: input url dtls://127.0.0.1:4444.
Successfully parsed a group of options.
Opening an input file: dtls://127.0.0.1:4444.
[AVFormatContext @ 0x616000000380] Opening 'dtls://127.0.0.1:4444' for reading
[dtls @ 0x60d000002b50] No default whitelist set
[udp @ 0x60c000021f40] No default whitelist set
[udp @ 0x60c000021f40] end receive buffer size reported is 393216
[dtls @ 0x616000000680] Info method=SSL_accept state=PINIT(before SSL initialization), where=8193, ret=1
    Last message repeated 1 times
[dtls @ 0x616000000680] Info method=SSL_accept state=TRCH(SSLv3/TLS read client hello), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TWSH(SSLv3/TLS write server hello), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TWSC(SSLv3/TLS write certificate), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TWSKE(SSLv3/TLS write key exchange), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TWCR(SSLv3/TLS write certificate request), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TWSD(SSLv3/TLS write server done), where=8193, ret=1
    Last message repeated 1 times
[dtls @ 0x616000000680] Info method=SSL_accept state=TRCC(SSLv3/TLS read client certificate), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TRCKE(SSLv3/TLS read client key exchange), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TRCV(SSLv3/TLS read certificate verify), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TRCCS(SSLv3/TLS read change cipher spec), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TRFIN(SSLv3/TLS read finished), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TWST(SSLv3/TLS write session ticket), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TWCCS(SSLv3/TLS write change cipher spec), where=8193, ret=1
[dtls @ 0x616000000680] Info method=SSL_accept state=TWFIN(SSLv3/TLS write finished), where=8193, ret=1
[dtls @ 0x616000000680] Setup ok, MTU=1500
[mpegts @ 0x616000000380] Format mpegts probed with size=2048 and score=50
[mpegts @ 0x616000000380] stream=0 stream_type=1b pid=100 prog_reg_desc=
[mpegts @ 0x616000000380] Before avformat_find_stream_info() pos: 0 bytes read:3384 seeks:0 nb_streams:1
[h264 @ 0x618000000c80] Decoding VUI
[extract_extradata @ 0x6080000022a0] nal_unit_type: 9(AUD), nal_ref_idc: 0
[extract_extradata @ 0x6080000022a0] nal_unit_type: 7(SPS), nal_ref_idc: 3
[extract_extradata @ 0x6080000022a0] nal_unit_type: 8(PPS), nal_ref_idc: 3
[extract_extradata @ 0x6080000022a0] nal_unit_type: 6(SEI), nal_ref_idc: 0
[extract_extradata @ 0x6080000022a0] nal_unit_type: 5(IDR), nal_ref_idc: 3
[h264 @ 0x618000000c80] nal_unit_type: 9(AUD), nal_ref_idc: 0
[h264 @ 0x618000000c80] nal_unit_type: 7(SPS), nal_ref_idc: 3
[h264 @ 0x618000000c80] nal_unit_type: 8(PPS), nal_ref_idc: 3
[h264 @ 0x618000000c80] nal_unit_type: 6(SEI), nal_ref_idc: 0
[h264 @ 0x618000000c80] nal_unit_type: 5(IDR), nal_ref_idc: 3
[h264 @ 0x618000000c80] Decoding VUI
[h264 @ 0x618000000c80] Format yuv420p chosen by get_format().
[h264 @ 0x618000000c80] Reinit context to 128x80, pix_fmt: yuv420p
[h264 @ 0x618000000c80] no picture 
[mpegts @ 0x616000000380] max_analyze_duration 5000000 reached at 5000000 microseconds st:0
[mpegts @ 0x616000000380] After avformat_find_stream_info() pos: 112424 bytes read:112800 seeks:0 frames:152
Input #0, mpegts, from 'dtls://127.0.0.1:4444':
  Duration: N/A, start: 1.466667, bitrate: N/A
  Program 1 
    Metadata:
      service_name    : Service01
      service_provider: FFmpeg
  Stream #0:0[0x100], 152, 1/90000: Video: h264 (High), 1 reference frame ([27][0][0][0] / 0x001B), yuv420p(progressive, left), 128x72 [SAR 1:1 DAR 16:9], 0/1, 30 fps, 30 tbr, 90k tbn, start 1.466667
Successfully opened the file.
Parsing a group of options: output url out.mp4.
Applying option c (select encoder/decoder ('copy' to copy stream without reencoding)) with argument copy.
Successfully parsed a group of options.
Opening an output file: out.mp4.
[out#0/mp4 @ 0x611000034ac0] No explicit maps, mapping streams automatically...
[vost#0:0/copy @ 0x617000005f00] Created video stream from input stream 0:0
[file @ 0x60c000033f40] Setting default whitelist 'file,crypto,data'
Successfully opened the file.
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
Output #0, mp4, to 'out.mp4':
  Metadata:
    encoder         : Lavf62.1.102
  Stream #0:0, 0, 1/90000: Video: h264 (High), 1 reference frame (avc1 / 0x31637661), yuv420p(progressive, left), 128x72 [SAR 1:1 DAR 16:9], 0/1, q=2-31, 30 fps, 30 tbr, 90k tbn
[out#0/mp4 @ 0x611000034ac0] Starting thread...
[in#0/mpegts @ 0x6120000001c0] Starting thread...
Press [q] to stop, [?] for help
frame=  169 fps=0.0 q=-1.0 size=       0KiB time=00:00:05.56 bitrate=   0.1kbits/s speed=11.1x elapsed=0:0frame=  185 fps=184 q=-1.0 size=       0KiB time=00:00:06.10 bitrate=   0.1kbits/s speed=6.06x elapsed=0:0frame=  200 fps=133 q=-1.0 size=       0KiB time=00:00:06.60 bitrate=   0.1kbits/s speed=4.37x elapsed=0:0frame=  216 fps=107 q=-1.0 size=       0KiB time=00:00:07.13 bitrate=   0.1kbits/s speed=3.55x elapsed=0:0frame=  232 fps= 92 q=-1.0 size=       0KiB time=00:00:07.66 bitrate=   0.1kbits/s speed=3.05x elapsed=0:0frame=  248 fps= 82 q=-1.0 size=       0KiB time=00:00:08.20 bitrate=   0.0kbits/s speed=2.72x elapsed=0:0[h264 @ 0x618000008c80] Decoding VUI
frame=  298 fps= 85 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed= 2.8x elapsed=0:0frame=  298 fps= 74 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=2.45x elapsed=0:0frame=  298 fps= 66 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=2.18x elapsed=0:0frame=  298 fps= 59 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=1.96x elapsed=0:0frame=  298 fps= 54 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=1.78x elapsed=0:0frame=  298 fps= 49 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=1.63x elapsed=0:0frame=  298 fps= 46 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=1.51x elapsed=0:0frame=  298 fps= 42 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed= 1.4x elapsed=0:0frame=  298 fps= 39 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=1.31x elapsed=0:0frame=  298 fps= 37 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=1.22x elapsed=0:0frame=  298 fps= 35 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=1.15x elapsed=0:0frame=  298 fps= 33 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=1.09x elapsed=0:0frame=  298 fps= 31 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=1.03x elapsed=0:0frame=  298 fps= 30 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=0.98x elapsed=0:0frame=  298 fps= 28 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=0.933x elapsed=0:frame=  298 fps= 27 q=-1.0 size=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=0.891x elapsed=0:^C[in#0/mpegts @ 0x6120000001c0] Terminating thread with return code 0 (success)
[out#0/mp4 @ 0x611000034ac0] All streams finished
[out#0/mp4 @ 0x611000034ac0] Terminating thread with return code 0 (success)
[out#0/mp4 @ 0x611000034ac0] Error writing trailer: Immediate exit requested
[AVIOContext @ 0x612000004fc0] Statistics: 48 bytes written, 2 seeks, 4 writeouts
[out#0/mp4 @ 0x611000034ac0] Error closing file: Immediate exit requested
[out#0/mp4 @ 0x611000034ac0] Output file #0 (out.mp4):
[out#0/mp4 @ 0x611000034ac0]   Output stream #0:0 (video): 298 packets muxed (139588 bytes); 
[out#0/mp4 @ 0x611000034ac0]   Total: 298 packets (139588 bytes) muxed
[out#0/mp4 @ 0x611000034ac0] video:136KiB audio:0KiB subtitle:0KiB other streams:0KiB global headers:0KiB muxing overhead: unknown
frame=  298 fps= 24 q=-1.0 Lsize=       0KiB time=00:00:09.86 bitrate=   0.0kbits/s speed=0.789x elapsed=0:00:12.50    
[in#0/mpegts @ 0x6120000001c0] Input file #0 (dtls://127.0.0.1:4444):
[in#0/mpegts @ 0x6120000001c0]   Input stream #0:0 (video): 299 packets read (139889 bytes); 
[in#0/mpegts @ 0x6120000001c0]   Total: 299 packets (139889 bytes) demuxed
[AVIOContext @ 0x612000003340] Statistics: 218644 bytes read, 0 seeks
Exiting normally, received signal 2.

Exiting with exit code 255
```

I'm not sure whether it is an issue from DTLS?

But I noticed there're many dtls code need to improve, such as DTLS_method, cert and key init and so on.

I plan to work on implementing WHIP DTLS client role(only server role support for now) next step since I already submit the NACK patchset.
So If you could mentoring me something, I'm glad to fix such code!

> BTW, I can't join in IRC cause of my network environment, so I left this message to you.

Thanks!
Jack

JackLau1222 commented:
> Hi Timo,
> 
> I noticed you have tried that test DTLS using ffmpeg command but meet error, I test in my local using such command(need base on this udp patch):


I found the root reason is DTLS haven't properly free some variable
Already fix in https://github.com/BtbN/FFmpeg/pull/3/commits/59cb100079cd6daf2d5103e0b5d6fd624b43fd48

Now you can use the command and works well with openssl DTLS.
```shell
ffmpeg -v debug -listen 1 -cert_file cert.pem -key_file key.pem -mtu 1500 -autodetect_dest 1 -i dtls://127.0.0.1:4444 -c copy -y out.mp4

ffmpeg -v trace -f lavfi -re -i testsrc2=duration=10:size=128x72:rate=30 -an -sn -c:v libx264 -cert_file cert.pem -key_file key.pem -mtu 10000 -f mpegts "dtls://127.0.0.1:4444"
```

BtbN commented:
The patch set is merged, so you should be able to base on master again. 

When trying those commands, for me it's the client side that fails completely. 
The handshake just fails without ever even trying to send anything, just returning error 5, and print error got nothing in the error queue.

JackLau1222 commented:
> When trying those commands, for me it's the client side that fails completely. The handshake just fails without ever even trying to send anything, just returning error 5, and print error got nothing in the error queue.

I think it's related to that MTU didn't set in your command, and current MTU default value is 0
```c
int ffurl_write2(void *urlcontext, const uint8_t *buf, int size)
{
    URLContext *h = urlcontext;

    if (!(h->flags & AVIO_FLAG_WRITE))
        return AVERROR(EIO);
    /* avoid sending too big packets */
    if (h->max_packet_size && size > h->max_packet_size)
        return AVERROR(EIO);

    return retry_transfer_wrapper(h, NULL, buf, size, size, 0);
}
```
In that situation, h->max_packet_size will be 0, if the size > 0, it will return AVERROR(EIO) (-5).

BtbN commented:
The bio write method is never called though, so that can't be it.

JackLau1222 commented:
> The bio write method is never called though, so that can't be it.

I noticed you have set ret value when call `ff_udp_set_remote_addr`, so return ret=0 on the last, it is wrong because ret should be return code of `ffurl_read(c->tls_shared.is_dtls ? c->tls_shared.udp : c->tls_shared.tcp, buf, len);` is data length.

So it will not be try to call bwrite since openssl already meet an error.

I fix that in my code, and remove the modify of udp, I think you're right about that, it should be caller's work rather than let the underlay protocol do all thing

Right now we can just use such command to work well with openssl dtls:
```shell
ffmpeg -v debug -listen 1 -cert_file cert.pem -key_file key.pem -mtu 1500 -i dtls://127.0.0.1:4444 -c copy -y out.mp4
ffmpeg -v trace -f lavfi -re -i testsrc2=duration=5:size=128x72:rate=30 -an -sn -c:v libx264 -mtu 10000 -f mpegts "dtls://127.0.0.1:4444"
```

Could you try my latest code and give me some reviews?

Thanks!

JackLau1222 commented:
I totally agree your ideas about the handshake can just use loop, but openssl dtls handshake don't need a loop since we don't set it as NONBLOCK mode, in that situation, openssl will do the loop internally so one call just can be finished.

And when it set to NONBLOCK mode(just like WHIP did), it need be called multi-times, we can let the user(WHIP) to achevie that.

In a word, I don't really think we do the handshake loop in tls_openssl is meaningful, because it just need called once when it's not NONBLOCK mode.
What do you think about it?

BtbN commented:
Yep, very much did mix up the return values there.

Accumulated a lot of smaller and larger patches again now. Latest set is on master here and on the ML.

BtbN commented on libavformat/tls_openssl.c:
```diff
@@ -707,9 +707,6 @@ static int dtls_handshake(URLContext *h)
 
-    int was_nonblock = h->flags & AVIO_FLAG_NONBLOCK;
-    h->flags &= ~AVIO_FLAG_NONBLOCK;
-
```
SSL_do_handshake will call the bread and bwrite BIO callbacks, which very much DO look at this flag.
They will then forward the flag to the udp/tcp context, and thus the read/write will return EAGAIN, breaking the handshake, if this is not set here.

JackLau1222 replied:
the bread and bwrite using the udp or tcp context, isn't?
```C
static int url_bio_bwrite(BIO *b, const char *buf, int len)
{
    TLSContext *c = BIO_get_data(b);
    int ret = ffurl_write(c->tls_shared.is_dtls ? c->tls_shared.udp : c->tls_shared.tcp, buf, len);
    if (ret >= 0)
        return ret;
    BIO_clear_retry_flags(b);
    if (ret == AVERROR_EXIT)
        return 0;
    if (ret == AVERROR(EAGAIN))
        BIO_set_retry_write(b);
    else
        c->io_err = ret;
    return -1;
}
```
I don't see anywhere the dtls->flags pass into udp, it seems should manually pass just like this:
```C
static int tls_write(URLContext *h, const uint8_t *buf, int size)
{
    TLSContext *c = h->priv_data;
    URLContext *uc = c->tls_shared.is_dtls ? c->tls_shared.udp
                                           : c->tls_shared.tcp;
    int ret;

    // Set or clear the AVIO_FLAG_NONBLOCK on c->tls_shared.tcp
    uc->flags &= ~AVIO_FLAG_NONBLOCK;
    uc->flags |= h->flags & AVIO_FLAG_NONBLOCK;

    if (c->tls_shared.is_dtls)
        size = FFMIN(size, DTLS_get_data_mtu(c->ssl));

    ret = SSL_write(c->ssl, buf, size);
    if (ret > 0)
        return ret;
    if (ret == 0)
        return AVERROR_EOF;
    return print_ssl_error(h, ret);
}
```

And I debug into udp code and ensure this handshake code didn't pass flags into udp

BtbN replied:
Ah, you're right. It's the tls_write and tls_read functions doing this, not internally in the BIO.
It's setting the wrong flag then, it should set it on the udp URLContext (and also doesn't need to remember the old state).
Just plain removing this entirely seems wrong, the dlts_open function would need to loop again then, which could potentially be a busy loop.
I really don't see a sensible way to do the handshake in a non-blocking way. You always end up busy looping on the handshake function.

JackLau1222 replied:
I totally agree we default set it BLOCK mode, then the openssl will internally loop so we just call the dtls_handshake once.

But I think maybe we should give the user the ability to decide whether set NONBLOCK mode or not?(Just like WHIP does)

What do you think?

Thanks

BtbN replied:
What is the user to do though? There is no sensible mode of operation to do this handshake in a nonblocking way.
You always just end up looping over this do_handshake function, in what is potentially a busy loop, unless you build a very ugly construct that extracts the underlying fd, and calls select on it.

Keep in mind that whip is not the only user of the tls/dtls contexts (even though in practice it so far is, but that's the wrong way to think about that code).
If someone were to open a dtls:// url and set the nonblocking flag on open, it would be very surprising for them to then still have to loop over the handshake function until it succeeds, and do find some way to do it in a way that's not a busy loop.
Sure, they could late-set the flag, but that's just silly. In what situation would you ever want or need the handshake to not block?

JackLau1222 replied:
The NONBLCOK mode has advantages in webrtc, we maybe need to handle retransmission or other situations that require fine control. We might need this feature in future.

BtbN replied:
It doesn't mean that nonblocking reads/writes are impossible. This is only the initial handshake.

JackLau1222 replied:
Yes, the openssl also can retransmission internally.

So we could use BLOCK mode firstly.

I'll submit this patchset.

Thanks

BtbN commented:
I just merged the majority of the patches here, if you rebase onto master, not much should be left.

JackLau1222 commented:
> I just merged the majority of the patches here, if you rebase onto master, not much should be left.

please apply this also
https://github.com/BtbN/FFmpeg/pull/3/commits/fd7c2a85e46172fc410a89961530b68aadb19b06

Thanks!

BtbN commented:
I still had comments about that one.
Mostly that it's perfectly valid to av_free(p) a NULL pointer, so most of the if() in that patch are wholly unnecessary.
I'm not sure if the same is true for ffurl_closep(), so if it is, that if can also just be removed.

BtbN commented:
Also, aren't cert_buf and key_buf string AVOptions? Those don't need to be manually freed in general.
Or is there any situation where the code allocates them itself, without them having been passed in as an option?

JackLau1222 commented:
> Also, aren't cert_buf and key_buf string AVOptions? Those don't need to be manually freed in general. Or is there any situation where the code allocates them itself, without them having been passed in as an option?

There is no code allocates them, just pass by AVOptions, Thanks for your reminder! I'll fix that soon

BtbN commented on libavformat/tls_openssl.c:
```diff
@@ -725,6 +726,8 @@ static int dtls_handshake(URLContext *h)
     p->tls_shared.state = DTLS_STATE_FINISHED;
 end:
+    if (was_nonblock)
+        c->udp->flags |= AVIO_FLAG_NONBLOCK;
```
This isn't necessary, since tls_write and tls_read reset this flag from the TLSContext every single time.

JackLau1222 replied:
how about the latest code?

I reset udp to NONBLOCK in whip rtp so that we don't need modify dtls code

JackLau1222 committed:
avformat/whip: force set NONBLOCK for rtp

we need to ensure the rtp set NONBLOCK since dtls handshake has override to BLOCK mode

Signed-off-by: Jack Lau <jacklau1222@qq.com>

BtbN commented on libavformat/whip.c:
```diff
@@ -1467,6 +1467,9 @@ static int on_rtp_write_packet(void *opaque, const uint8_t *buf, int buf_size)
 
+    if (!(whip->udp->flags & AVIO_FLAG_NONBLOCK)) 
+        whip->udp->flags |= AVIO_FLAG_NONBLOCK;
+
```
Does whip actually still directly write to the UDP socket after it has handed it over to the DTLS URLContext?
Does that not break the DTLS session?
In that case, yeah, it does need to set the nonblock flag to the flags it desires.

On that note, the if is entirely unnecessary. It's going to take more CPU cycles to check if the flag is set than it takes to simply set it. Though of course we're talking negligible amounts of cycles here.

JackLau1222 replied:
WHIP reuse the same udp socket by ICE, DTLS, SRTP. 
So yes, whip write to the udp socket after the dtls handshake 
And it won't break dtls session because we actually use the SRTP to transfer data rather than dtls. we don't use dtls anymore after it is finished.

I can set the flag just once after dtls done. please see the latest code.

BtbN commented on libavformat/whip.c:
```diff
@@ -1496,6 +1496,9 @@ static int create_rtp_muxer(AVFormatContext *s)
 
+    if (!(whip->udp->flags & AVIO_FLAG_NONBLOCK)) 
+        whip->udp->flags |= AVIO_FLAG_NONBLOCK;
+
```
Either place is fine, what I meant is that the if() itself here is not needed.
Simply `whip->udp->flags |= AVIO_FLAG_NONBLOCK;` will do it, since it's faster than the check if it's unset itself.

JackLau1222 replied:
Sure, you're right, already fix that, thanks!

BtbN commented:
That should be all patches from here merged. So closing.
